src/ModelContextProtocol/Protocol/Transport/InMemoryTransport.cs

```diff
@@ -0,0 +1,416 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+using ModelContextProtocol.Configuration;
+using ModelContextProtocol.Logging;
+using ModelContextProtocol.Protocol.Messages;
+using ModelContextProtocol.Server;
+
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Channels;
+
+
+namespace ModelContextProtocol.Protocol.Transport;
+
+/// <summary>
+/// Provides an in memory implementation of the MCP transport protocol over shared memory.
+/// This transport enables efficient in-process MCP functionality and easier test authoring.
+/// </summary>
+/// <remarks>
+/// <para>
+/// The InMemoryTransport allows both client and server to communicate within the same process,
+/// which is particularly useful for testing scenarios and embedded applications.
+/// </para>
+/// <para>
+/// This implementation requires dynamic code access for tool registration and might not work in Native AOT.
+/// </para>
+/// </remarks>
+[RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
+public sealed class InMemoryTransport : TransportBase, IServerTransport, IClientTransport
```

Comment:
I'm unclear what the intended usage of this is. It contains only a single channel, so presumably this can't be used for both client and server. Is the intent that two separate transports are created, one for client to send to server and one for server to send to client? I'd like to see a test in this PR that actually uses InMemoryTransport with McpClientFactory and McpServerFactory to establish a session.

Reply:
still working on this was a draft, needs 2 channels

Reply:
I want to have duplex communication.